src/Common/src/TypeSystem/Common/CastingHelper.cs

```diff
@@ -0,0 +1,380 @@
+                            && thisArrayType.CanCastParamTo(otherArrayType.ParameterType, protect);
+
+                    case TypeKind.DefType:
+                        return thisType.CanCastToClassOrInterface(otherType, protect);
+                }
+            }
+
+            return false;
+        }
+
+        private static bool CanCastParamTo(this ParameterizedType thisType, TypeDesc paramType, StackOverflowProtect protect)
+        {
+            // While boxed value classes inherit from object their
+            // unboxed versions do not.  Parameterized types have the
+            // unboxed version, thus, if the from type parameter is value
+            // class then only an exact match/equivalence works.
+            if (thisType.ParameterType == paramType)
+            {
+                return true;
+            }
+
+            return ParametrizedTypeCastHelper(thisType.ParameterType, paramType, protect);
+        }
+
+        private static bool IsObjRef(this TypeFlags flags)
+        {
+            return flags == TypeFlags.Class || flags == TypeFlags.Array;
+        }
+        
+        private static bool ParametrizedTypeCastHelper(TypeDesc curTypesParm, TypeDesc otherTypesParam, StackOverflowProtect protect)
+        {
+            // Object parameters don't need an exact match but only inheritance, check for that
+            TypeDesc fromParamUnderlyingType = curTypesParm.UnderlyingType;
+            if (fromParamUnderlyingType.Category.IsObjRef())
+            {
+                return curTypesParm.CanCastToInternal(otherTypesParam, protect);
+            }
+            else if (curTypesParm.IsGenericParameter)
+            {
+                var genericVariableFromParam = (GenericParameterDesc)curTypesParm;
+                if (genericVariableFromParam.HasReferenceTypeConstraint)
+                {
+                    return genericVariableFromParam.CanCastToInternal(otherTypesParam, protect);
+                }
+            }
+            else if (fromParamUnderlyingType.IsPrimitive)
+            {
+                TypeDesc toParamUnderlyingType = otherTypesParam.UnderlyingType;
+                if (toParamUnderlyingType.IsPrimitive)
+                {
+                    if (toParamUnderlyingType == fromParamUnderlyingType)
+                    {
+                        return true;
+                    }
+
+                    // Primitive types such as E_T_I4 and E_T_U4 are interchangeable
+                    // Enums with interchangeable underlying types are interchangable
+                    // BOOL is NOT interchangeable with I1/U1, neither CHAR -- with I2/U2
+                    // Float and dobule are not interchangable here.
+                    TypeFlags fromParamCategory = fromParamUnderlyingType.Category;
+                    TypeFlags toParamCategory = toParamUnderlyingType.Category;
+
+                    if (fromParamCategory != TypeFlags.Boolean &&
```

Comment:
The other cast implementations do this by comparing normalized sizes (https://github.com/dotnet/corert/blob/4376c55dd018d8b7b9fed82449728711231ba266/src/Runtime.Base/src/System/Runtime/TypeCast.cs#L951). I think it is easier to understand. Either way - it would be nice to have this in separate method.